Updates for cold start half cycle, then continuing with IAU for WCDA

JessicaMeixner-NOAA opened the pull request:
<!-- PLEASE READ -->
<!-- Any PRs not following this template will be closed -->
<!--
    Please use a short (<60 char), descriptive title above. It should complete the sentence "If merged, this PR will _____". Capitalize the first word and do not end with a period.

    PRs should meet these guidelines:
    - Each PR should address ONE topic and have an associated issue.
    - No hard-coded paths or personal directories.
    - No temporary or backup files should be committed (including logs).
    - Any code that you disabled by being commented out should be removed or reenabled.

    Please delete all these comments before submitting the PR.
-->
# Description
<!-- This description will become the commit message for the PR-->
<!-- Please use this format for your description:

  Describe your changes. Focus on the *what* and *why*. The *how* will be evident from the changes. In particular, be sure to note any interface changes, such as command line syntax, that will need to be communicated to users.

  At the end of your description, please be sure to add the issue this PR solves using the word "Resolves". If there are any issues that are related but not yet resolved (including in other repos), you may use "Refs".

  Resolves #1234
  Refs #4321
  Refs NOAA-EMC/repo#5678
-->

This PR allows us to run C384 S2S with IAU, but starting with the first half-cycle as a cold-start.   This will be necessary for cycled testing as we build towards the full system for GFSv17.  This updates the copying of the restarts for RUN=gdas for both ocean and ice copying what the atm model is doing.  It also reduced the amount of restart files from 4 to 3. 

Refs #2546

Other updates: 
* Add DOJEDI ocean triggers for archiving certain files update from:  @CatherineThomas-NOAA 
* Adds COPY_FINAL_RESTARTS option to turn on/off copying the last restart file to COM.  Defaults to off... 
* Defines model_start_date_current_cycle & model_start_date_next_cycle to help with knowing which IC to grab. 

# Type of change
<!-- Delete all except one -->
- New feature (adds functionality)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
<!-- Please list any test you conducted, including the machine.

Example:
- Clone and build on WCOSS
- Cycled test on Orion
- Forecast-only on Hera
-->

This was tested by running C384 WCDA test on hera which can be found here: 
/scratch1/NCEPDEV/climate/Jessica.Meixner/cycling/iau_04/C384iaucold01 
which includes yaml case files and then in cold01 has COM/EXPDIRs

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] I have made corresponding changes to the documentation if necessary

JessicaMeixner-NOAA committed:
update gdas app hash and turn on C48mx500_3DVarAOWCDA test on hera

JessicaMeixner-NOAA committed:
Bug fix for #2528 from aerorahul

JessicaMeixner-NOAA committed:
add additional update from before

JessicaMeixner-NOAA committed:
remove line that made things "work" but for the wrong reasons

JessicaMeixner-NOAA committed:
updates to run with 3 instead of 4 mom restarts
updates to have correct ICs in com for IAU after a cold start first half cycle

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/c384wcda

JessicaMeixner-NOAA committed:
update CICE restart copying

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/c384wcda

JessicaMeixner-NOAA committed:
update gdas app hash

JessicaMeixner-NOAA committed:
update gdas

JessicaMeixner-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/c384wcda

JessicaMeixner-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/c384wcda

JessicaMeixner-NOAA commented:
So I still think this PR should move forward as is, but I'm going to change resolve to ref for the issue, because I've noticed that the gfsfcst is actually failing when copying restarts after the forecast is over.  The scripts are expecting a forecast hour that MOM6 is not producing.  That issue is tracked in #1776

JessicaMeixner-NOAA committed:
add additional archive switch for ocean/ice DA files for if JEDIOCN is included
--- fixes from Cathy

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/c384wcda' into feature/c384wcda

aerorahul commented on ush/forecast_postdet.sh:
```diff
@@ -500,22 +500,20 @@ MOM6_out() {
-      restart_file="${restart_date:0:8}.${restart_date:8:2}0000.${mom6_restart_file}"
-      ${NCP} "${DATArestart}/MOM6_RESTART/${restart_file}" \
-             "${COM_OCEAN_RESTART}/${restart_file}"
+    restart_date=$(date --utc -d "${current_cycle:0:8} ${current_cycle:8:2} + ${restart_interval} hours" +%Y%m%d%H)
+    while (( restart_date < forecast_end_cycle )); do
+      echo "Copying MOM6 restarts for 'RUN=${RUN}' at ${restart_date}"
+      for mom6_restart_file in "${mom6_restart_files[@]}"; do
+	restart_file="${restart_date:0:8}.${restart_date:8:2}0000.${mom6_restart_file}"
+        ${NCP} "${DATArestart}/MOM6_RESTART/${restart_file}" \
+               "${COM_OCEAN_RESTART}/${restart_file}"
+      done
+      restart_date=$(date --utc -d "${restart_date:0:8} ${restart_date:8:2} + ${restart_interval} hours" +%Y%m%d%H)
```
If I understand this correctly, `COM` will contain 2 pairs of restarts; one at the beginning and another at the end of the assimilation window regardless of `DOIAU`.
Is there a way to separate that, such that only the neccessary restart is copied out to `COM`.
If `DOIAU=NO`, it will always be the middle of the window.
If `DOIAU=YES`, it will always be the beginning of the window.

I can see the previous logic was deficient in handling `DOIAU=YES` for the experiment that has a cold-start half cycle.

JessicaMeixner-NOAA replied:
I'm not sure what another solution would be.  Is there a variable that exists that indicates that it's the first half cycle?  What is here is using what the atmosphere model is doing for gdas, but if there's another solution I can implement that here instead.

aerorahul replied:
JEDI-based atmosphere DA requires restarts in the assimilation window, and that is why they are being copied to `COM`. 

A variable to indicate the first half cycle is a good idea to explore.

JessicaMeixner-NOAA replied:
What about: 

https://github.com/NOAA-EMC/global-workflow/blob/develop/ush/forecast_det.sh#L30-L32 These lines: 
```
    # Since warm start is false, we cannot do IAU
    DOIAU="NO"
    IAU_OFFSET=0
```
to 
```
    # Since warm start is false, we cannot do IAU - but set a variable to know the next cycle will have IAU
    if [[ "${DOIAU}" == "YES" ]]; then 
          NEXTCYCIAU=YES
    else 
          NEXTCYCIAU=NO 
    fi 
    DOIAU="NO"
    IAU_OFFSET=0
```
And then we can use "NEXTCYCIAU" in the copying to know which restart to copy?

JessicaMeixner-NOAA replied:
I slightly updated this, but am using a "NEXTCYCIAU trigger for which restarts to copy for CICE, MOM6 and CMEPS (someday can be used for ww3 too).    This would reduce the amount of restarts copied.  I'll push updates after testing.  I've marked the PR as draft to resolve this issue.

JessicaMeixner-NOAA replied:
@aerorahul trying something to see if it works.  Let me know what you think, we can try something else.  This was inspired by your ideas you shared offline, but might not have been exactly what you were originally thinking so we can tweak as needed.

JessicaMeixner-NOAA committed:
tab to spaces

JessicaMeixner-NOAA committed:
add COPY_FINAL_RESTARTS flag to skip copying restarts at the end

JessicaMeixner-NOAA committed:
saving work with NEXTCYC date, likely will use different solution

JessicaMeixner-NOAA committed:
updating logic for DOIAU related date changes in fewer places
removing extra DOIAU if logic in multiple places... still needs testing

JessicaMeixner-NOAA commented on ush/forecast_predet.sh:
```diff
@@ -52,9 +52,13 @@ common_predet(){
-  current_cycle_end=$(date --utc -d "${current_cycle:0:8} ${current_cycle:8:2} + ${half_window} hours" +%Y%m%d%H)
-  next_cycle_begin=$(date --utc -d "${next_cycle:0:8} ${next_cycle:8:2} - ${half_window} hours" +%Y%m%d%H)
-  next_cycle_end=$(date --utc -d "${next_cycle:0:8} ${next_cycle:8:2} + ${half_window} hours" +%Y%m%d%H)
+  if [[ "${DOIAU:-}" == "YES" ]]; then
```
@aerorahul my test is still going, but this seems to be working as expected with this fix.  If current_cycle_begin and next_cycle_begin should not change definitions, we could add a variable here perhaps startdate_current_cycle startdate_next_cycle that could be defined and updated as current_cycle_begin/next_cycle_begin are being defined here?

aerorahul replied:
I am working on a patch that will:
1. have the ability to control if we want to save last time restarts
2. save restarts at the appropriate time of next cycle when doing IAU, but the experiments start  with a cold start

JessicaMeixner-NOAA replied:
What are the issues with the solution (or alternative) shared here?

aerorahul replied:
I don't believe this is correct.
Regardless of DOIAU, current_cycle_begin is the time of the beginning of the assimilation window. If it makes it more clear to call this current_cycle_window_begin, then we can rename w/ `window` in the word.
current_cycle and next_cycle are CDATE and CDATE+6h.  Those are fixed.

JessicaMeixner-NOAA replied:
So what I understand is that you would like for the variables  current_cycle_begin to not be changed.   In this case, I can add the new variables: startdate_current_cycle startdate_next_cycle or a name you would prefer that would allow for current_cycle_begin/next_cycle_begin to remain unchanged.    Is startdate_current_cycle startdate_next_cycle  acceptable names?

JessicaMeixner-NOAA replied:
I've made an update that hopefully addresses your concerns.  Please let me know if there are other outstanding concerns I can address.

aerorahul replied:
Yes. Thank you.   
`sdate_current_cycle` is the `model_start_date_current_cycle` and `sdate_next_cycle` is the `model_start_date_next_cycle`. And this is captured here rather than the many places where `DOIAU` is relevant.

JessicaMeixner-NOAA replied:
Awesome!  Would you prefer the names be model_start_date_current_cycle and model_start_date_next_cycle? If so I can update that on Monday!

aerorahul replied:
I will leave it to you; a descriptive name is useful.

JessicaMeixner-NOAA replied:
I updated the name so it's more descriptive.

aerorahul replied:
Thanks for cleaning this up.  It looks great.

JessicaMeixner-NOAA committed:
create two new variables that define the start time
of the forecast model for current and next cycle and leave other
defintions of variables as is

aerorahul commented on ush/forecast_predet.sh:
```diff
@@ -54,7 +54,15 @@ common_predet(){
+  forecast_end_cycle=$(date --utc -d "${current_cycle:0:8} ${current_cycle:8:2} + ${FHMAX} hours" +%Y%m%d%H)
+  #Define sdate for current_cycle and next_cycle as the time the forecast will start for this cycle and next cycle
+  if [[ "${DOIAU:-}" == "YES" ]]; then
+    sdate_current_cycle=$(date --utc -d "${current_cycle:0:8} ${current_cycle:8:2} - ${half_window} hours" +%Y%m%d%H)
+    sdate_next_cycle=$(date --utc -d "${next_cycle:0:8} ${next_cycle:8:2} - ${half_window} hours" +%Y%m%d%H)
```
These are calculated on the few lines above and it makes casts `model_start_date_current_cycle` and `model_start_date_next_cycle` in relation to the anchors above.
```suggestion
    sdate_current_cycle="${current_cycle_begin}"
    sdate_next_cycle="${next_cycle_begin}"
```

JessicaMeixner-NOAA committed:
Update ush/forecast_predet.sh

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/c384wcda

JessicaMeixner-NOAA committed:
use more descriptive variable name sdate -> model start date

aerorahul commented on ush/forecast_predet.sh:
```diff
@@ -54,7 +54,15 @@ common_predet(){
   current_cycle_end=$(date --utc -d "${current_cycle:0:8} ${current_cycle:8:2} + ${half_window} hours" +%Y%m%d%H)
   next_cycle_begin=$(date --utc -d "${next_cycle:0:8} ${next_cycle:8:2} - ${half_window} hours" +%Y%m%d%H)
-  next_cycle_end=$(date --utc -d "${next_cycle:0:8} ${next_cycle:8:2} + ${half_window} hours" +%Y%m%d%H)
+  forecast_end_cycle=$(date --utc -d "${current_cycle:0:8} ${current_cycle:8:2} + ${FHMAX} hours" +%Y%m%d%H)
```
```suggestion
```
Repeated on line 66.

aerorahul reviewed (approved):
This looks great. Thank you!

JessicaMeixner-NOAA committed:
update cice output ic link

JessicaMeixner-NOAA committed:
update remove data command - suggestion from Rahul

JessicaMeixner-NOAA commented:
@aerorahul I have marked this as ready to review.  We'll do a separate PR to add in some updates @guillaumevernieres is working which is one reason this was still review.

JessicaMeixner-NOAA committed:
Update ush/forecast_predet.sh

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

JessicaMeixner-NOAA committed:
add new parm variable for gefs

JessicaMeixner-NOAA commented:
@aerorahul I just updated the description and remembered that i hadn't added the variable for GEFS, so also added that.

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/c384wcda

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/c384wcda

aerorahul commented:
@NeilBarton-NOAA 
Some of these updates will be relevant for the PR #2559

aerorahul commented on scripts/exglobal_archive.sh:
```diff
@@ -237,7 +237,10 @@ if [[ ${HPSSARCH} = "YES" || ${LOCALARCH} = "YES" ]]; then
         if [ "${DO_OCN}" = "YES" ]; then
-            targrp_list="${targrp_list} gdasocean gdasocean_analysis"
+            targrp_list="${targrp_list} gdasocean"	  
+            if [ "${DO_JEDIOCNVAR}" = "YES" ]; then
```
```suggestion
            if [[ "${DO_JEDIOCNVAR}" == "YES" ]]; then
```

aerorahul replied:
This should resolve the shellnorm failure.

JessicaMeixner-NOAA committed:
Update scripts/exglobal_archive.sh

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

JessicaMeixner-NOAA commented:
@aerorahul I just wanted to chime in here to say that my test of the  C384mx025 3DVarAOWCDA with a fresh clone is running smoothly as well.

guillaumevernieres reviewed (approved):
Thanks @JessicaMeixner-NOAA . Tested over a couple of days of cycling.
